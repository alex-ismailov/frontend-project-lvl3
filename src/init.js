/* eslint no-param-reassign: 0 */

import i18next from 'i18next';
import * as yup from 'yup';
import onChange from 'on-change';
import axios from 'axios';
import _ from 'lodash';
import resources from './locales/index.js';
import {
  renderInputError, renderFeeds, renderFeedback, renderPosts, addDataToModal, renderViewedPost,
} from './view.js';
import parse from './parser.js';

const TIMEOUT = 5000; // ms
const DELAY = 5000; // ms
const processStateMap = {
  filling: 'filling',
  sending: 'sending',
  finished: 'finished',
  failed: 'failed',
};

const buildAllOriginsUrl = (rssUrl) => {
  const corsProxy = 'https://hexlet-allorigins.herokuapp.com';
  const corsProxyApi = '/get';
  const params = new URLSearchParams();
  params.append('disableCache', true);
  params.append('url', rssUrl);

  return `${corsProxy}${corsProxyApi}?${params.toString()}`;
};

const addNewRssFeed = (watchedState) => {
  const { form: { value: feedUrl } } = watchedState;
  axios.get(buildAllOriginsUrl(feedUrl), { timeout: TIMEOUT })
    .then((response) => {
      const rawData = response.data.contents;
      if (!rawData.startsWith('<?xml')) {
        throw new Error('notValidRssFormat');
      }
      const feedData = parse(rawData, feedUrl);
      watchedState.feeds = [feedData.feedInfo, ...watchedState.feeds];
      watchedState.posts = [...feedData.posts, ...watchedState.posts];
      watchedState.form.valid = true;
      watchedState.form.error = '';
      watchedState.form.value = '';
      watchedState.form.processState = processStateMap.finished;
      watchedState.form.processState = processStateMap.filling;
    })
    .catch((e) => {
      const message = e.message === 'notValidRssFormat'
        ? i18next.t('errors.notValidRssFormat')
        : i18next.t('errors.networkError');
      watchedState.form.error = message;
      watchedState.form.processState = processStateMap.failed;
    });
};

const watchForNewPosts = (watchedState, timerId) => {
  clearTimeout(timerId);
  const { feeds } = watchedState;
  const promises = feeds.map(({ link }) => axios.get(buildAllOriginsUrl(link), { timeout: TIMEOUT })
    .then((v) => ({ result: 'success', value: v, feedUrl: link }))
    .catch((e) => ({ result: 'error', error: e, feedUrl: link })));
  const promise = Promise.all(promises);

  return promise.then((responses) => {
    const freshPosts = responses.flatMap((response) => {
      if (response.result === 'error') {
        // TODO: надо как-то обработать этот кейс
        // console.log(`Impossible to get data from: ${response.feedUrl}`);
        return [];
      }
      const { value, feedUrl } = response;
      const feedData = parse(value.data.contents, feedUrl);
      return feedData.posts;
    });

    const newPosts = _.differenceBy(freshPosts, watchedState.posts, 'title');
    if (!_.isEmpty(newPosts)) {
      watchedState.posts = [...newPosts, ...watchedState.posts];
    }

    const newTimerId = setTimeout(() => watchForNewPosts(watchedState, newTimerId), DELAY);
  });
};

// *** MODEL ***
export default () => {
  i18next.init({
    fallbackLng: 'ru',
    resources,
  });

  const state = {
    form: {
      processState: processStateMap.filling,
      valid: true,
      value: '',
      error: '',
    },
    feeds: [],
    posts: [],
    uiState: {
      modal: {
        currentPostId: null,
      },
      currentViewedPostId: null,
      viewedPostsIds: new Set(),
    },
  };

  const form = document.getElementById('rssForm');
  const input = form.elements.url;
  input.focus();
  const feedback = document.querySelector('.feedback');
  const submitButton = form.querySelector('button');
  const feedsBlock = document.querySelector('.feeds');
  const postsBlock = document.querySelector('.posts');

  const handleProcessState = (processState) => {
    switch (processState) {
      case 'filling':
        submitButton.disabled = false;
        break;
      case 'failed':
        submitButton.disabled = false;
        input.readOnly = false;
        input.focus();
        break;
      case 'sending':
        submitButton.disabled = true;
        input.readOnly = true;
        break;
      case 'finished':
        renderFeedback('success', feedback);
        submitButton.disabled = false;
        input.readOnly = false;
        input.value = '';
        input.focus();
        break;
      default:
        throw new Error(`Unknown process state: ${processState}`);
    }
  };

  const watchedState = onChange(state, (path, value) => {
    switch (path) {
      case 'form.processState':
        handleProcessState(value);
        break;
      case 'form.valid':
        renderInputError(value, input);
        break;
      case 'form.error':
        renderFeedback(value, feedback);
        break;
      case 'feeds':
        renderFeeds(value, feedsBlock);
        break;
      case 'posts':
        /*  я прокидываю watchedState через view, потому что
        во время рендеринга постов renderPosts динмачески создает
        новые контроллеры для кнопок preview, которые в свою очередь тоже
        должны как-то иметь доступ к модели, чтобы устанавливать id
        текущего активнога поста для модального окна.
        Не уверен можно ли прокидывать модель через view для динам.
        создаваемого контроллера. */
        renderPosts(value, watchedState, postsBlock);
        break;
      case 'uiState.modal.currentPostId': {
        const post = watchedState.posts.find(({ id }) => id === value);
        addDataToModal(post);
        break;
      }
      case 'uiState.currentViewedPostId':
        renderViewedPost(value);
        break;
      default:
        break;
    }
  });

  // *** VIEW ***
  // look at src/js/view.js
  // ************

  const schema = yup
    .string()
    .url(i18next.t('errors.notValidUrl'))
    .test('unique', i18next.t('errors.feedExists'),
      (value) => !watchedState.feeds.some((feed) => feed.link === value));

  const validate = (value) => {
    try {
      schema.validateSync(value, { abortEarly: false });
      return null;
    } catch (e) {
      return e.message;
    }
  };

  // *** CONTROLLERS ***
  form.addEventListener('input', (e) => {
    const { target: { value } } = e;
    watchedState.form.value = value;
  });

  form.addEventListener('submit', (e) => {
    e.preventDefault();
    const error = validate(watchedState.form.value);
    if (error) {
      watchedState.form.processState = processStateMap.failed;
      watchedState.form.valid = false;
      watchedState.form.error = error;
      return;
    }
    addNewRssFeed(watchedState);
    watchedState.form.processState = processStateMap.sending;
  });

  // контроллер демон watchForNewPosts, запускается один раз на этапе инициализации приложения
  const timerId = setTimeout(() => watchForNewPosts(watchedState, timerId), DELAY);
};
