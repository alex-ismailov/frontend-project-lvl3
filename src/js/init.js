/* eslint no-param-reassign: 0 */

import axios from 'axios';
import onChange from 'on-change';
import * as yup from 'yup';
import {
  renderInputError, renderError, renderFeeds, renderPosts,
} from './view.js';
import parse from './parser.js';

// *** MVC: MODEL -> VIEW -> CONTROLLER ->> MODEL ......***
// ********************************************************

// *** utils ***
const schema = yup.string().required().url('Must be valid url');

const validate = (watchedState) => {
  const { form: { value }, feeds } = watchedState;
  try {
    schema.validateSync(value, { abortEarly: false });
    return feeds.some((feed) => value.includes(feed.link))
      ? 'Rss already exists'
      : '';
  } catch (e) {
    return e.message;
  }
};

const addNewRssFeed = (url, watchedState) => {
  axios.get(url, { timeout: 1000 })
    .then((response) => {
      if (response.data.status.http_code === 404) {
        watchedState.form.error = 'This source doesn\'t contain valid rss';
        watchedState.form.processState = 'failed';
        return;
      }
      const feedData = parse(response);
      watchedState.feeds = [feedData.feed, ...watchedState.feeds];
      watchedState.posts = [...feedData.posts, ...watchedState.posts];

      watchedState.form.processState = 'finished';
      watchedState.form.valid = true;
      watchedState.form.error = '';
      watchedState.form.value = '';
      watchedState.form.processState = 'filling';
    })
    .catch((e) => {
      watchedState.form.error = e.message;
      watchedState.form.processState = 'failed';
    });
};

// *** VIEW ***
// look at src/js/view.js
// ************

// **********************************************************************
export default () => {
  console.log('Start !!!!');
  // имена состояний это причастия: ....
  // *** MODEL ***
  const state = {
    form: {
      processState: 'filling', // sending, finished || failed
      valid: true,
      value: '',
      error: '',
    },
    feeds: [],
    posts: [],
  };

  const form = document.getElementById('rssForm');
  const input = form.elements.url;
  const feedback = document.querySelector('.feedback');
  const submitButton = form.querySelector('button');
  input.focus();

  const processStateHandler = (processState) => {
    switch (processState) {
      case 'filling':
        submitButton.disabled = false;
        break;
      case 'failed':
        submitButton.disabled = false;
        input.disabled = false;
        input.focus();
        break;
      case 'sending':
        submitButton.disabled = true;
        input.disabled = true;
        break;
      case 'finished':
        submitButton.disabled = false;
        input.disabled = false;
        input.value = '';
        input.focus();
        break;
      default:
        throw new Error(`Unknown process state: ${processState}`);
    }
  };

  const watchedState = onChange(state, (path, value) => {
    switch (path) {
      case 'form.processState':
        processStateHandler(value);
        break;
      case 'form.valid':
        renderInputError(value, input);
        break;
      case 'form.error':
        renderError(value, feedback);
        break;
      case 'feeds': {
        console.log(value);
        renderFeeds();
        break;
      }
      case 'posts': {
        console.log(value);
        renderPosts();
        break;
      }
      default:
        break;
    }
  });

  // *** CONTROLLERS ***
  form.addEventListener('input', (e) => {
    const { target: { value } } = e;
    watchedState.form.value = value;
  });

  form.addEventListener('submit', (e) => {
    e.preventDefault();
    const error = validate(watchedState);
    if (error) {
      watchedState.form.processState = 'failed';
      watchedState.form.valid = false;
      watchedState.form.error = error;
      return;
    }
    const urlWithAllOriginsProxy = `https://hexlet-allorigins.herokuapp.com/get?url=${encodeURIComponent(watchedState.form.value)}`;
    addNewRssFeed(urlWithAllOriginsProxy, watchedState);
    watchedState.form.processState = 'sending';
  });
};
