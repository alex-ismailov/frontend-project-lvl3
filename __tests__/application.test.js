import '@testing-library/jest-dom';

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
// import { screen, waitFor } from '@testing-library/dom';
// import userEvent from '@testing-library/user-event';
import testingLibraryDom from '@testing-library/dom';
import testingLibraryUserEvent from '@testing-library/user-event';
import nock from 'nock';

import init from '../src/js/init.js';

const { screen } = testingLibraryDom;
const userEvent = testingLibraryUserEvent.default;

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const getFixturePath = (filename) => path.join(__dirname, '..', '__fixtures__', filename);
const readFixture = (filename) => {
  const fixturePath = getFixturePath(filename);

  const rss = fs.readFileSync(fixturePath, 'utf-8');
  return rss;
};
const rss1 = readFixture('rss1.xml');
// const rss2 = readFixture('rss2.xml');
// const rss3 = readFixture('rss3.xml');
const rssUrl = 'https://ru.hexlet.io/lessons.rss';
const corsProxy = 'https://hexlet-allorigins.herokuapp.com';
const corsProxyApi = '/get';

// const htmlPath = getFixturePath('document.html');
// const html = fs.readFileSync(htmlPath, 'utf-8');
// const htmlUrl = 'https://ru.hexlet.io';

const index = path.join(__dirname, '..', 'index.html');
const initHtml = fs.readFileSync(index, 'utf-8');

const elements = {};

beforeAll(() => {
  nock.disableNetConnect();
});

afterAll(() => {
  nock.cleanAll();
  nock.enableNetConnect();
});

beforeEach(async () => {
  document.body.innerHTML = initHtml;

  await init();

  elements.input = screen.getByRole('textbox', { name: 'url' });
  elements.submit = screen.getByRole('button', { name: 'add' });
});

test('adding', async () => {
  nock(corsProxy)
    .defaultReplyHeaders({
      'access-control-allow-origin': '*',
      'access-control-allow-credentials': 'true',
    })
    .get(corsProxyApi)
    .query({ url: rssUrl, disableCache: 'true' })
    .reply(200, { contents: rss1 });

  userEvent.type(elements.input, rssUrl);
  userEvent.click(elements.submit);
  expect(await screen.findByText(/RSS успешно загружен/i)).toBeInTheDocument();
});


test('validation url', async () => {
  userEvent.type(elements.input, 'bad url');
  userEvent.click(elements.submit);
  expect(screen.getByText(/Ссылка должна быть валидным URL/i)).toBeInTheDocument();
});

test('validation unique url', async () => {
  nock(corsProxy)
    .defaultReplyHeaders({
      'access-control-allow-origin': '*',
      'access-control-allow-credentials': 'true',
    })
    .get(corsProxyApi)
    .query({ url: rssUrl, disableCache: 'true' })
    .reply(200, { contents: rss1 });

  userEvent.type(elements.input, rssUrl);
  userEvent.click(elements.submit);
  expect(await screen.findByText(/RSS успешно загружен/i)).toBeInTheDocument();
  
  userEvent.type(elements.input, rssUrl);
  userEvent.click(elements.submit);
  expect(await screen.findByText(/RSS уже существует/i)).toBeInTheDocument();
});
