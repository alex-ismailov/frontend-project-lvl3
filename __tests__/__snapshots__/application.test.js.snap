// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#addNewFeedWithTwoPosts1 1`] = `
<body>
  


  
  <meta
    charset="UTF-8"
  />
  
  
  <meta
    content="width=device-width, initial-scale=1.0"
    name="viewport"
  />
  
  
  <title>
    RSS Aggregator
  </title>
  


  
  <div
    aria-hidden="true"
    aria-labelledby="modal"
    class="modal fade"
    id="modal"
    role="dialog"
    tabindex="-1"
  >
    
    
    <div
      class="modal-dialog"
      role="document"
    >
      
      
      <div
        class="modal-content"
      >
        
        
        <div
          class="modal-header"
        >
          
          
          <h5
            class="modal-title"
          />
          
          
          <button
            aria-label="Close"
            class="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
          
        
        </div>
        
        
        <div
          class="modal-body"
        />
        
        
        <div
          class="modal-footer"
        >
          
          
          <a
            class="btn btn-primary full-article"
            rel="noopener noreferrer"
            role="button"
            target="_blank"
          >
            Full article
          </a>
          
          
          <button
            class="btn btn-secondary"
            data-bs-dismiss="modal"
            type="button"
          >
            Close
          </button>
          
        
        </div>
        
      
      </div>
      
    
    </div>
    
  
  </div>
  
  
  <main
    class="flex-grow-1"
  >
    
    
    <section
      class="input-form container-fluid bg-dark p-5"
    >
      
      
      <div
        class="row"
      >
        
        
        <div
          class="col-md-10 col-lg-8 mx-auto text-white"
        >
          
          
          <h1
            class="input-form__title display-3"
          >
            RSS Reader
          </h1>
          
          
          <p
            class="lead"
          >
            Start reading RSS today! It is easy, it is nicely.
          </p>
          
          
          <form
            action=""
            class="rss-form"
            id="rssForm"
          >
            
            
            <div
              class="row"
            >
              
              
              <div
                class="col-12 col-sm"
              >
                
                
                <input
                  class="form-control form-control-lg w-100"
                  disabled=""
                  id="rssFormInput"
                  name="url"
                  placeholder="RSS link"
                  required=""
                  type="text"
                />
                
              
              </div>
              
              
              <div
                class="col-12 col-sm-auto mt-3 mt-sm-0"
              >
                
                
                <button
                  class="btn btn-lg btn-primary px-sm-5 w-100 w-sm-auto"
                  disabled=""
                  type="submit"
                >
                  Add
                </button>
                
              
              </div>
              
            
            </div>
            
          
          </form>
          
          
          <p
            class="text-muted my-1"
          >
            Example: https://ru.hexlet.io/lessons.rss
          </p>
          
          
          <div
            class="feedback"
          />
          
        
        </div>
        
      
      </div>
      
    
    </section>
    
    
    <section
      class="container-fluid p-5"
    >
      
      
      <div
        class="row"
      >
        
        
        <div
          class="col-md-10 col-lg-8 mx-auto feeds"
        />
        
      
      </div>
      
      
      <div
        class="row"
      >
        
        
        <div
          class="col-md-10 col-lg-8 mx-auto posts"
        />
        
      
      </div>
      
    
    </section>
    
  
  </main>
  
  
  <footer
    class="footer border-top py-3 mt-5"
  >
    
    
    <div
      class="container-xl"
    >
      
      
      <div
        class="text-center"
      >
        
        created by
        
        <a
          class="text-decoration-none"
          href="https://github.com/alex-ismailov"
          target="_blank"
        >
          Alex Ismailov
        </a>
        
      
      </div>
      
    
    </div>
    
  
  </footer>
  


</body>
`;

exports[`#freshApplication1 1`] = `
"<meta charset=\\"UTF-8\\" />
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
<title>RSS Aggregator</title>

<div
  class=\\"modal fade\\"
  id=\\"modal\\"
  tabindex=\\"-1\\"
  role=\\"dialog\\"
  aria-labelledby=\\"modal\\"
  aria-hidden=\\"true\\"
>
  <div class=\\"modal-dialog\\" role=\\"document\\">
    <div class=\\"modal-content\\">
      <div class=\\"modal-header\\">
        <h5 class=\\"modal-title\\"></h5>
        <button
          type=\\"button\\"
          class=\\"btn-close\\"
          data-bs-dismiss=\\"modal\\"
          aria-label=\\"Close\\"
        ></button>
      </div>
      <div class=\\"modal-body\\"></div>
      <div class=\\"modal-footer\\">
        <a
          class=\\"btn btn-primary full-article\\"
          role=\\"button\\"
          target=\\"_blank\\"
          rel=\\"noopener noreferrer\\"
        >
          Full article
        </a>
        <button type=\\"button\\" class=\\"btn btn-secondary\\" data-bs-dismiss=\\"modal\\">
          Close
        </button>
      </div>
    </div>
  </div>
</div>
<main class=\\"flex-grow-1\\">
  <section class=\\"input-form container-fluid bg-dark p-5\\">
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto text-white\\">
        <h1 class=\\"input-form__title display-3\\">RSS Reader</h1>
        <p class=\\"lead\\">Start reading RSS today! It is easy, it is nicely.</p>
        <form id=\\"rssForm\\" action=\\"\\" class=\\"rss-form\\">
          <div class=\\"row\\">
            <div class=\\"col-12 col-sm\\">
              <input
                required=\\"\\"
                id=\\"rssFormInput\\"
                type=\\"text\\"
                name=\\"url\\"
                class=\\"form-control form-control-lg w-100\\"
                placeholder=\\"RSS link\\"
              />
            </div>
            <div class=\\"col-12 col-sm-auto mt-3 mt-sm-0\\">
              <button
                type=\\"submit\\"
                class=\\"btn btn-lg btn-primary px-sm-5 w-100 w-sm-auto\\"
              >
                Add
              </button>
            </div>
          </div>
        </form>
        <p class=\\"text-muted my-1\\">Example: https://ru.hexlet.io/lessons.rss</p>
        <div class=\\"feedback\\"></div>
      </div>
    </div>
  </section>
  <section class=\\"container-fluid p-5\\">
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto feeds\\"></div>
    </div>
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto posts\\"></div>
    </div>
  </section>
</main>
<footer class=\\"footer border-top py-3 mt-5\\">
  <div class=\\"container-xl\\">
    <div class=\\"text-center\\">
      created by
      <a
        class=\\"text-decoration-none\\"
        href=\\"https://github.com/alex-ismailov\\"
        target=\\"_blank\\"
      >
        Alex Ismailov
      </a>
    </div>
  </div>
</footer>
"
`;
